File: e2e/tests/steps/media_state.rs
use std::time::Duration;

use cucumber::{given, then, when};
use medea_e2e::object::{
    remote_track::MediaDirection, AwaitCompletion, MediaSourceKind,
};

use crate::World;

use super::{parse_media_kind, parse_media_kinds};

#[given(regex = r"^(\S+)'s `getUserMedia\(\)` request has added latency$")]
async fn given_gum_delay(world: &mut World, id: String) {
    let member = world.get_member(&id).unwrap();
    member.add_gum_latency(Duration::from_millis(500)).await;
}

#[when(regex = r"^(\S+) frees all local tracks$")]
async fn when_member_frees_all_local_tracks(world: &mut World, id: String) {
    let member = world.get_member(&id).unwrap();
    member.room().forget_local_tracks().await;
}

#[then(regex = "^(\\S+)'s (audio|(?:device|display) video) local track is \
                 (not )?muted$")]
async fn then_local_track_mute_state(
    world: &mut World,
    id: String,
    kind: String,
    not_muted: String,
) {
    let member = world.get_member(&id).unwrap();
    let (media_kind, source_kind) = parse_media_kinds(&kind).unwrap();
    let track = member
        .room()
        .local_tracks()
        .await
        .unwrap()
        .get_track(media_kind, source_kind)
        .await
        .unwrap();
    assert_eq!(not_muted.is_empty(), track.muted().await.unwrap());
}

#[then(regex = "^(\\S+)'s (audio|(?:device|display) video) local track is \
                 stopped$")]
async fn then_track_is_stopped(world: &mut World, id: String, kind: String) {
    let member = world.get_member(&id).unwrap();
    let (media_kind, source_kind) = parse_media_kinds(&kind).unwrap();
    let is_stopped = member
        .room()
        .local_tracks()
        .await
        .unwrap()
        .get_track(media_kind, source_kind)
        .await
        .unwrap()
        .free_and_check()
        .await
        .unwrap();
    assert!(is_stopped);
}

#[then(regex = "^(\\S+)'s (audio|video) from (\\S+) has \
             (SendRecv|SendOnly|RecvOnly|Inactive) direction$")]
async fn then_remote_media_direction_is(
    world: &mut World,
    id: String,
    kind: String,
    remote_id: String,
    direction: String,
) {
    let media_kind = kind.parse().unwrap();
    let media_direction = match direction.as_str() {
        "SendRecv" => MediaDirection::SendRecv,
        "SendOnly" => MediaDirection::SendOnly,
        "RecvOnly" => MediaDirection::RecvOnly,
        "Inactive" => MediaDirection::Inactive,
        _ => unreachable!(),
    };

    let member = world.get_member(&id).unwrap();
    let connection = member
        .connections()
        .wait_for_connection(remote_id)
        .await
        .unwrap();
    let tracks_store = connection.tracks_store().await.unwrap();
    let track = tracks_store
        .get_track(media_kind, MediaSourceKind::Device)
        .await
        .unwrap();
    track
        .wait_for_media_direction(media_direction)
        .await
        .unwrap();
}

#[when(regex = "^(\\S+) (enables|disables|mutes|unmutes) (audio|video)\
                 ( and awaits it (complete|error)s)?$")]
async fn when_enables_or_mutes(
    world: &mut World,
    id: String,
    action: String,
    audio_or_video: String,
    awaits: String,
) {
    let member = world.get_member(&id).unwrap();
    let maybe_await = if awaits.is_empty() {
        AwaitCompletion::Dont
    } else {
        AwaitCompletion::Do
    };

    let result = match action.as_str() {
        "enables" => {
            member
                .toggle_media(
                    parse_media_kind(&audio_or_video),
                    None,
                    true,
                    maybe_await,
                )
                .await
        }
        "disables" => {
            member
                .toggle_media(
                    parse_media_kind(&audio_or_video),
                    None,
                    false,
                    maybe_await,
                )
                .await
        }
        "mutes" => {
            member
                .toggle_mute(
                    parse_media_kind(&audio_or_video),
                    None,
                    true,
                    maybe_await,
                )
                .await
        }
        _ => {
            member
                .toggle_mute(
                    parse_media_kind(&audio_or_video),
                    None,
                    false,
                    maybe_await,
                )
                .await
        }
    };

    if maybe_await == AwaitCompletion::Do {
        if awaits.contains("error") {
            result.unwrap_err();
        } else {
            result.unwrap();
        }
    }
}

#[when(regex = "^(\\S+) switches device with latency$")]
async fn when_member_switches_device_with_latency(
    world: &mut World,
    id: String,
) {
    let member = world.get_member(&id).unwrap();

    member.add_gum_latency(Duration::from_secs(3)).await;
    member.switch_video_device().await.unwrap();
}

#[when(regex = "^(\\S+) (enables|disables) remote \
                 (audio|(?:device |display )?video)$")]
async fn when_member_enables_remote_track(
    world: &mut World,
    id: String,
    toggle: String,
    kind: String,
) {
    let member = world.get_member(&id).unwrap();
    let media_kind = kind.parse().unwrap();
    let source_kind = kind.parse().ok();

    if toggle == "enables" {
        member
            .room()
            .enable_remote_media(media_kind, source_kind)
            .await
            .unwrap();
    } else {
        member
            .room()
            .disable_remote_media(media_kind, source_kind)
            .await
            .unwrap();
    }
}

File: flutter/test/e2e/steps/media_state.dart
import 'dart:async';

import 'package:flutter_test/flutter_test.dart';
import 'package:gherkin/gherkin.dart';

import 'package:medea_jason/medea_jason.dart';
import '../world/custom_world.dart';
import '../world/more_args.dart';

List<StepDefinitionGeneric> steps() {
  return [
    givenGumDelay,
    thenLocalTrackMuteState,
    thenRemoteMediaDirectionIs,
    thenTrackIsStopped,
    whenEnablesOrMutes,
    whenMemberEnablesRemoteTrack,
    whenMemberFreesAllLocalTracks,
    whenMemberSwitchesDeviceWithLatency
  ];
}

StepDefinitionGeneric whenEnablesOrMutes =
    when4<String, String, String, String, CustomWorld>(
  RegExp(r'(\S+) (enables|disables|mutes|unmutes) (audio|video) and'
      r' (awaits it completes|awaits it errors|ignores the result)?$'),
  (id, action, audioOrVideo, String awaits, context) async {
    var kind = parseMediaKind(audioOrVideo);
    var member = context.world.members[id]!;

    var awaitable = awaits.contains('awaits');
    var error = awaits.contains('errors');
    Future<void> future;
    switch (action) {
      case 'enables':
        future = member.toggleMedia(kind.item1, kind.item2, true);
        break;

      case 'disables':
        future = member.toggleMedia(kind.item1, kind.item2, false);
        break;

      case 'mutes':
        future = member.toggleMute(kind.item1, kind.item2, true);
        break;

      default:
        future = member.toggleMute(kind.item1, kind.item2, false);
        break;
    }

    if (awaitable) {
      try {
        await future;
      } catch (e) {
        if (!error) {
          rethrow;
        }
      }
    } else {
      unawaited(future.catchError((e) => {
            // suppress error
          }));
    }
  },
);

StepDefinitionGeneric whenMemberEnablesRemoteTrack =
    when3<String, String, String, CustomWorld>(
  RegExp(r'(\S+) (enables|disables) remote '
      r'(audio|(?:device |display )?video)$'),
  (id, toggle, String kind, context) async {
    var parsedKind = parseMediaKind(kind);
    var member = context.world.members[id]!;

    if (toggle == 'enables') {
      if (parsedKind.item1 == MediaKind.audio) {
        await member.room.enableRemoteAudio();
      } else {
        await member.room.enableRemoteVideo();
      }
    } else {
      if (parsedKind.item1 == MediaKind.audio) {
        await member.room.disableRemoteAudio();
      } else {
        await member.room.disableRemoteVideo();
      }
    }
  },
);

StepDefinitionGeneric thenRemoteMediaDirectionIs =
    then4<String, String, String, String, CustomWorld>(
  RegExp(r"(\S+)'s (audio|video) from (\S+) has "
      r'`(SendRecv|SendOnly|RecvOnly|Inactive)` direction$'),
  (id, String kind, remoteId, direction, context) async {
    var member = context.world.members[id]!;

    var parsedKind = parseMediaKind(kind);

    await member.waitForConnect(remoteId);
    var track = await member.waitRemoteTrackFrom(
        remoteId, parsedKind.item2, parsedKind.item1);

    var dir = TrackMediaDirection.values
        .firstWhere((e) => e.name.toLowerCase() == direction.toLowerCase());

    await member.waitMediaDirectionTrack(dir, track);
  },
);

StepDefinitionGeneric thenLocalTrackMuteState =
    then3<String, String, String, CustomWorld>(
  RegExp(r"(\S+)'s (audio|(?:device|display) video) local track is "
      r'(not )?muted$'),
  (id, String kind, notMuted, context) async {
    var member = context.world.members[id]!;
    var parsedKind = parseMediaKind(kind);

    var track = await member.waitLocalTrack(parsedKind.item2, parsedKind.item1);
    var muted = !notMuted.contains('not');
    expect(!track.getTrack().isEnabled(), muted);
  },
);

StepDefinitionGeneric thenTrackIsStopped = then2<String, String, CustomWorld>(
  RegExp(r"(\S+)'s (audio|(?:device|display) video) local track is "
      r'stopped$'),
  (id, kind, context) async {
    var member = context.world.members[id]!;
    var parsedKind = parseMediaKind(kind);
    var track = await member.waitLocalTrack(parsedKind.item2, parsedKind.item1);

    await track.free();
  },
);

StepDefinitionGeneric whenMemberFreesAllLocalTracks =
    when1<String, CustomWorld>(
  RegExp(r'(\S+) frees all local tracks$'),
  (id, context) async {
    var member = context.world.members[id]!;
    await member.forgetLocalTracks();
  },
);

StepDefinitionGeneric whenMemberSwitchesDeviceWithLatency =
    when1<String, CustomWorld>(
  RegExp(r'(\S+) switches device with latency$'),
  (id, context) async {
    var member = context.world.members[id]!;
    member.setGumLatency(const Duration(seconds: 3));
    await member.switchVideoDevice();
  },
);

StepDefinitionGeneric givenGumDelay = given1<String, CustomWorld>(
  RegExp(r"(\S+)'s `getUserMedia\(\)` request has added latency$"),
  (id, context) async {
    var member = context.world.members[id]!;
    member.setGumLatency(const Duration(milliseconds: 500));
  },
);
